### PR TITLE
perf: improve FlagDependencyExportsPlugin for large JSON by depth

### PR DESCRIPTION
**Pull request by hai-x**
## Summary
Aligned with https://github.com/webpack/webpack/pull/19058.

I will update doc later.

## Checklist

- [x] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by hai-x** [comment]
parser.json.exportsDepth: 1
image

parser.json.exportsDepth: Infinity
image

File: packages/rspack/src/config/defaults.ts [hunk 1]
```diff
@@ -307,6 +310,14 @@ const applyModuleDefaults = (
 	assertNotNill(module.parser.javascript);
 	applyJavascriptParserOptionsDefaults(module.parser.javascript);
 
+	F(module.parser, JSON_MODULE_TYPE, () => ({}));
+	assertNotNill(module.parser[JSON_MODULE_TYPE]);
+	D(
+		module.parser[JSON_MODULE_TYPE],
+		"exportsDepth",
+		mode === "development" ? 1 : Number.POSITIVE_INFINITY
```
Comment (ahabhgk) [inline packages/rspack/src/config/defaults.ts:318]:
```suggestion
		mode === "development" ? 1 : Number.MAX_SAFE_INTEGER
```

**Review by ahabhgk — COMMENTED** [review]
Generally looks good to me

File: packages/rspack/src/config/defaults.ts (hunk 1 above)
Comment (hai-x) [reply packages/rspack/src/config/defaults.ts:318]: It is `Infinity` in webpack. Should we align?

File: crates/rspack_plugin_json/src/json_exports_dependency.rs [hunk 2]
```diff
@@ -13,13 +13,15 @@ pub struct JsonExportsDependency {
   id: DependencyId,
   #[cacheable(with=AsPreset)]
   data: JsonValue,
+  exports_depth: f64,
```
[the remark this answers is not in the document]
Comment (hai-x) [reply crates/rspack_plugin_json/src/json_exports_dependency.rs:16]: make sense.

File: packages/rspack/src/config/defaults.ts (hunk 1 above)
Comment (ahabhgk) [reply packages/rspack/src/config/defaults.ts:318]: I think both is fine, but I'm not sure what will pass to rust side by napi-rs if we use Infinity here

Comment (hai-x) [reply packages/rspack/src/config/defaults.ts:318]: Make sense. After tested napi locally. I Found that:

js pass `Infinity` & rust declare `f64` =>  exportsDepth: f64::INFINITY(rust side)
js pass `Infinity` & rust declare `u32` =>  exportsDepth: 0

js pass `MAX_SAFE_INTEGER` & rust declare `f64` =>  exportsDepth: 9007199254740991 (2^53-1)
js pass `MAX_SAFE_INTEGER` & rust declare `u32` =>  exportsDepth: 4294967295 (2^32-1)

Should we keep `f64` + `Infinity` to align?

Comment (ahabhgk) [reply packages/rspack/src/config/defaults.ts:318]: Both fine to me, but I prefer `u32` a little bit :)

Comment (hai-x) [reply packages/rspack/src/config/defaults.ts:318]: > Both fine to me, but I prefer `u32` a little bit :)

No problem. Let we use `u32`. I think it's more efficient in rust side. 

When there is feedback from users, we can make further changes.

I will update it later.

**Review by ahabhgk — APPROVED** [review]
Thanks!

**Comment by hai-x** [comment]
Update to `1.2.0-alpha.0` and found that

when `module.parser.json.exportsDepth: Number.MAX_SAFE_INTEGER`

```
$ rspack build --config dev.config.js
Rspack compiled successfully in 1.94 s
✨  Done in 2.94s.
```

when `module.parser.json.exportsDepth: Number.MAX_SAFE_INTEGER`

```
$ rspack build --config dev.config.js
Rspack compiled successfully in 505 ms
✨  Done in 1.34s.
```

**Pull request merged**